packages/kbn-babel-plugin-apm-idx/index.js

```diff
@@ -0,0 +1,251 @@
+/*
```

Comment:
Diff against the original work (https://github.com/facebookincubator/idx/blob/462c7ecd3ca717503c35606be0e7f355715ddf99/packages/babel-plugin-idx/src/babel-plugin-idx.js):
```
diff --git a/./babel-plugin-idx.js b/./kbn-babel-plugin-apm-idx.js
index 81f4f65..d37eb28 100644
--- a/./babel-plugin-idx.js
+++ b/./kbn-babel-plugin-apm-idx.js
@@ -1,3 +1,28 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * The following is the original license notice in the source
+ * code from which this code is based.
+ * (https://github.com/facebookincubator/idx/blob/462c7ecd3ca717503c35606be0e7f355715ddf99/packages/babel-plugin-idx/src/babel-plugin-idx.js)
+ */
+
 /**
  * Copyright (c) 2013-present, Facebook, Inc.
  *
@@ -7,8 +32,9 @@
  * @format
  */
 
-'use strict'; // eslint-disable-line strict
+/* eslint strict: 0, new-cap: 0 */
 
+'use strict';
 module.exports = context => {
   const t = context.types;
 
@@ -17,36 +43,33 @@ module.exports = context => {
   function checkIdxArguments(file, node) {
     const args = node.arguments;
     if (args.length !== 2) {
-      throw file.buildCodeFrameError(
-        node,
-        'The `idx` function takes exactly two arguments.',
-      );
+      throw file.buildCodeFrameError(node, 'The `idx` function takes exactly two arguments.');
     }
     const arrowFunction = args[1];
     if (!t.isArrowFunctionExpression(arrowFunction)) {
       throw file.buildCodeFrameError(
         arrowFunction,
-        'The second argument supplied to `idx` must be an arrow function.',
+        'The second argument supplied to `idx` must be an arrow function.'
       );
     }
     if (!t.isExpression(arrowFunction.body)) {
       throw file.buildCodeFrameError(
         arrowFunction.body,
         'The body of the arrow function supplied to `idx` must be a single ' +
-          'expression (without curly braces).',
+          'expression (without curly braces).'
       );
     }
     if (arrowFunction.params.length !== 1) {
       throw file.buildCodeFrameError(
         arrowFunction.params[2] || arrowFunction,
-        'The arrow function supplied to `idx` must take exactly one parameter.',
+        'The arrow function supplied to `idx` must take exactly one parameter.'
       );
     }
     const input = arrowFunction.params[0];
     if (!t.isIdentifier(input)) {
       throw file.buildCodeFrameError(
         arrowFunction.params[0],
-        'The parameter supplied to `idx` must be an identifier.',
+        'The parameter supplied to `idx` must be an identifier.'
       );
     }
   }
@@ -55,10 +78,7 @@ module.exports = context => {
     if (t.isImportDeclaration(node)) {
       // E.g. `import '...'`
       if (node.specifiers.length === 0) {
-        throw file.buildCodeFrameError(
-          node,
-          'The idx import must have a value.',
-        );
+        throw file.buildCodeFrameError(node, 'The idx import must have a value.');
       }
       // E.g. `import A, {B} from '...'`
       //      `import A, * as B from '...'`
@@ -66,16 +86,7 @@ module.exports = context => {
       if (node.specifiers.length > 1) {
         throw file.buildCodeFrameError(
           node.specifiers[1],
-          'The idx import must be a single specifier.',
-        );
-      }
-      // `import {default as idx} from '...'` or `import idx from '...'` are ok.
-      // `import idx, * as idx2 from '...'` is not ok but would've been caught
-      // above.
-      if (!t.isSpecifierDefault(node.specifiers[0])) {
-        throw file.buildCodeFrameError(
-          node.specifiers[0],
-          'The idx import must be a default import.',
+          'The idx import must be a single specifier.'
         );
       }
       // `importKind` is not a property unless flow syntax is enabled.
@@ -91,17 +102,14 @@ module.exports = context => {
         node.specifiers[0].importKind === 'type' ||
         node.specifiers[0].importKind === 'typeof'
       ) {
-        throw file.buildCodeFrameError(
-          node,
-          'The idx import must be a value import.',
-        );
+        throw file.buildCodeFrameError(node, 'The idx import must be a value import.');
       }
     } else if (t.isVariableDeclarator(node)) {
       // E.g. var {idx} or var [idx]
       if (!t.isIdentifier(node.id)) {
         throw file.buildCodeFrameError(
           node.specifiers[0],
-          'The idx declaration must be an identifier.',
+          'The idx declaration must be an identifier.'
         );
       }
     }
@@ -110,13 +118,9 @@ module.exports = context => {
   function makeCondition(node, state, inside) {
     if (inside) {
       return t.ConditionalExpression(
-        t.BinaryExpression(
-          '!=',
-          t.AssignmentExpression('=', state.temp, node),
-          t.NullLiteral(),
-        ),
+        t.BinaryExpression('!=', t.AssignmentExpression('=', state.temp, node), t.NullLiteral()),
         inside,
-        state.temp,
+        state.temp
       );
     } else {
       return node;
@@ -128,25 +132,21 @@ module.exports = context => {
       return makeChain(
         node.object,
         state,
-        makeCondition(
-          t.MemberExpression(state.temp, node.property, node.computed),
-          state,
-          inside,
-        ),
+        makeCondition(t.MemberExpression(state.temp, node.property, node.computed), state, inside)
       );
     } else if (t.isIdentifier(node)) {
       if (node.name !== state.base.name) {
         throw state.file.buildCodeFrameError(
           node,
           'The parameter of the arrow function supplied to `idx` must match ' +
-            'the base of the body expression.',
+            'the base of the body expression.'
         );
       }
       return makeCondition(state.input, state, inside);
     } else {
       throw state.file.buildCodeFrameError(
         node,
-        'idx callbacks may only access properties on the callback parameter.',
+        'idx callbacks may only access properties on the callback parameter.'
       );
     }
   }
@@ -163,21 +163,27 @@ module.exports = context => {
     });
     path.replaceWith(replacement);
     // Hoist to the top if it's an async method.
-    if (path.scope.path.isClassMethod({async: true})) {
-      path.scope.push({id: temp, _blockHoist: 3});
+    if (path.scope.path.isClassMethod({ async: true })) {
+      path.scope.push({ id: temp, _blockHoist: 3 });
     } else {
-      path.scope.push({id: temp});
+      path.scope.push({ id: temp });
     }
   }
 
   function isIdxImportOrRequire(node, name) {
     if (t.isImportDeclaration(node)) {
-      return t.isStringLiteral(node.source, {value: name});
+      if (name instanceof RegExp) {
+        return name.test(node.source.value);
+      } else {
+        return t.isStringLiteral(node.source, { value: name });
+      }
     } else if (t.isVariableDeclarator(node)) {
       return (
         t.isCallExpression(node.init) &&
-        t.isIdentifier(node.init.callee, {name: 'require'}) &&
-        t.isLiteral(node.init.arguments[0], {value: name})
+        t.isIdentifier(node.init.callee, { name: 'require' }) &&
+        (name instanceof RegExp
+          ? name.test(node.init.arguments[0].value)
+          : t.isLiteral(node.init.arguments[0], { value: name }))
       );
     } else {
       return false;
@@ -198,10 +204,7 @@ module.exports = context => {
       const idxBinding = path.scope.getOwnBinding(bindingName);
 
       idxBinding.constantViolations.forEach(refPath => {
-        throw state.file.buildCodeFrameError(
-          refPath.node,
-          '`idx` cannot be redefined.',
-        );
+        throw state.file.buildCodeFrameError(refPath.node, '`idx` cannot be redefined.');
       });
 
       let didTransform = false;
@@ -239,7 +242,7 @@ module.exports = context => {
           // "babel-plugin-transform-async-to-generator", will convert arrow
           // functions inside async functions into regular functions. So we do
           // our transformation before any one else interferes.
-          const newState = {file: state.file, importName};
+          const newState = { file: state.file, importName };
           path.traverse(declareVisitor, newState);
         }
       },
```